Enable Enable @ModelProp @TextProp and @CallbackProp with all options on field attribute

juliafu1 opened the pull request:
Enable Enable @ModelProp @TextProp and @CallbackProp with all options on field attribute

elihart commented on epoxy-processor/src/main/java/com/airbnb/epoxy/ViewAttributeInfo.kt:
```diff
@@ -9,31 +9,46 @@ import java.util.*
 private val NOT_NULL_ANNOTATION_SPEC = AnnotationSpec.builder(NotNull::class.java).build()
 private val NON_NULL_ANNOTATION_SPEC = AnnotationSpec.builder(NonNull::class.java).build()
 
+internal enum class ViewAttributeType {
```
can you make this a sealed class instead?

elihart commented on epoxy-processor/src/main/java/com/airbnb/epoxy/ViewAttributeInfo.kt:
```diff
@@ -9,31 +9,46 @@ import java.util.*
     val resetWithNull: Boolean
     val generateStringOverloads: Boolean
     var constantFieldNameForDefaultValue: String? = null
+    var viewAttributeTypeName: ViewAttributeType? = null
```
this should be a `val`

elihart commented on epoxy-processor/src/main/java/com/airbnb/epoxy/ViewAttributeInfo.kt:
```diff
@@ -333,7 +379,7 @@ internal class ViewAttributeInfo(
         }
 
-        builder.add("\n\n@see \$T#\$L(\$T)", viewElement.asType(), viewSetterMethodName, typeMirror)
+        builder.add("\n\n@see \$T#\$L(\$T)", viewElement.asType(), viewSetterPropName, typeMirror)
```
I don't think the generated javadoc will be right for fields

elihart commented on epoxy-processor/src/main/java/com/airbnb/epoxy/ViewAttributeInfo.kt:
```diff
@@ -272,6 +314,10 @@ internal class ViewAttributeInfo(
             typeName: TypeName
     ) {
+        //TODO juliafu support annotations for field
+        if (viewAttributeTypeName == ViewAttributeType.Field) {
```
I think this should just work already with no changes

elihart commented on epoxy-processor/src/main/java/com/airbnb/epoxy/ViewAttributeInfo.kt:
```diff
@@ -226,27 +261,34 @@ internal class ViewAttributeInfo(
+            setterElementOnView: Element
     ) {
+
+        if (!options.isEmpty() &&  viewAttributeTypeName == ViewAttributeType.Field) {
```
a field should be able to support options, I don't think it needs to be different there

elihart commented on epoxy-processor/src/main/java/com/airbnb/epoxy/ViewAttributeInfo.kt:
```diff
@@ -128,6 +152,17 @@ internal class ViewAttributeInfo(
+            element: Element,
+            errorLogger: ErrorLogger
+    ) {
+        when {
```
it's cleaner to return a value than to set a field. the field should be a val to be safer

elihart commented on epoxy-processor/src/main/java/com/airbnb/epoxy/ViewAttributeInfo.kt:
```diff
@@ -9,31 +9,46 @@ import java.util.*
 
         val options = HashSet<Option>()
-        val param = setterElementOnView.parameters[0]
+        val param = if (setterElementOnView is ExecutableElement) setterElementOnView.parameters[0]
```
`setterElementOnView` should be renamed, it doesn't make sense for fields

elihart commented on epoxy-processor/src/main/java/com/airbnb/epoxy/ViewAttributeInfo.kt:
```diff
@@ -43,6 +58,11 @@ internal class ViewAttributeInfo(
             options.addAll(Arrays.asList(*propAnnotation.value))
         } else if (textAnnotation != null) {
+            //TODO juliafu support textProp for field
+            if (viewAttributeTypeName == ViewAttributeType.Field) {
```
we shouldn't have to do anything special here to support it

elihart commented on epoxy-processor/src/main/java/com/airbnb/epoxy/ViewAttributeInfo.kt:
```diff
@@ -52,6 +72,10 @@ internal class ViewAttributeInfo(
             }
             options.add(Option.GenerateStringOverloads)
         } else if (callbackAnnotation != null) {
+            if (viewAttributeTypeName == ViewAttributeType.Field) {
```
same here

elihart commented on epoxy-processor/src/main/java/com/airbnb/epoxy/ViewAttributeInfo.kt:
```diff
@@ -75,8 +99,8 @@ internal class ViewAttributeInfo(
 
-        this.viewSetterMethodName = setterElementOnView.simpleName.toString()
-        propName = removeSetPrefix(viewSetterMethodName)
+        this.viewSetterPropName = setterElementOnView.simpleName.toString()
```
I don't know if the name `viewSetterPropName` makes sense for fields

juliafu1 commented:
@elihart I updated the code and addressed all comments
The testing is here: https://docs.google.com/document/d/1sT-AUi-Iiv3qI6dTfzbW3szRteq9qwaY-m1L7vnSY58/edit#bookmark=id.t7rle9y2y6xm

There are two things not fully working:
1) default value:
@ModelProp (defaultValue = "DEFAULT_STRING") String value;

The build stuck at  Task :epoxy-sample:compileDebugJavaWithJavac. I do not know why. There is no error message

2) annotation
@ModelProp @Nullable string value generates the following:

@Nullable
public TestViewModel_ value(final OnModelClickListener<TestViewModel_, TestView> value) {
  .....
}

I need to figure out how to make the @Nullable annotation on parameter. 

PTAL

elihart commented on epoxy-processor/src/main/java/com/airbnb/epoxy/ModelViewInfo.kt:
```diff
@@ -153,11 +153,12 @@ internal class ModelViewInfo(
                                        resourceProcessor))
     }
 
+    //TODO juliafu handle the super class case
```
can you add more details to this comment? not sure what needs to be done exactly

elihart commented on epoxy-processor/src/main/java/com/airbnb/epoxy/ModelViewProcessor.kt:
```diff
@@ -356,12 +380,14 @@ internal class ModelViewProcessor(
     companion object {
         val modelPropAnnotations = listOf(ModelProp::class, TextProp::class,
                                           CallbackProp::class).map { it.java }
+        val modelPropAnnotationsNames = listOf(ModelProp::class, TextProp::class,
```
it would be nice not to duplicate the list of prop classes

elihart commented on epoxy-processor/src/main/java/com/airbnb/epoxy/ModelViewWriter.kt:
```diff
@@ -214,9 +217,14 @@ internal class ModelViewWriter(
-                            viewAttribute.viewSetterMethodName,
-                            getValueToSetOnView(viewAttribute, boundObjectParam))
+        val expression = if (viewAttribute.viewAttributeTypeName == ViewAttributeType.Field)
+            "\$L.\$L=\$L;\n" else "\$L.\$L(\$L);\n"
```
I think we can do `CodeBlock.builder().addStatement()...` to avoid having to do `;\n` manually.

elihart replied:
this is mostly duplicated with the unbind code, the only difference is that the unbind code sets null. can you think of a way to DRY it out?

elihart commented on epoxy-processor/src/main/java/com/airbnb/epoxy/ViewAttributeInfo.kt:
```diff
@@ -128,6 +137,21 @@ internal class ViewAttributeInfo(
+    ): ViewAttributeType? {
+        return when {
+            element is ExecutableElement -> ViewAttributeType.Method
+            element.kind === ElementKind.FIELD -> ViewAttributeType.Field
```
don't use triple equals here

elihart replied:
can you do `when(element.kind) is FIELD ... is METHOD...`?

juliafu1 replied:
It couldn't use 'is'.  The compiler tells: "'is' over enum entry is not allowed, use comparison instead". I can use '==' instead of '===' but it actually doesn't make a difference for enum

elihart commented on epoxy-processor/src/main/java/com/airbnb/epoxy/ViewAttributeInfo.kt:
```diff
@@ -274,7 +298,10 @@ internal class ViewAttributeInfo(
+            if (ModelViewProcessor.modelPropAnnotationsNames.contains(ClassName.get(annotationElement as TypeElement))) {
+                continue
+            }
+            val builder = AnnotationSpec.builder(ClassName.get(annotationElement))
```
save `ClassName.get(annotationElement)` to a val since it is done twice?

elihart commented on epoxy-processor/src/main/java/com/airbnb/epoxy/ViewAttributeInfo.kt:
```diff
@@ -274,7 +298,10 @@ internal class ViewAttributeInfo(
         for (annotationMirror in paramElement.annotationMirrors) {
             val annotationElement = types.asElement(annotationMirror.annotationType)
-            val builder = AnnotationSpec.builder(ClassName.get(annotationElement as TypeElement))
+            if (ModelViewProcessor.modelPropAnnotationsNames.contains(ClassName.get(annotationElement as TypeElement))) {
```
can do `if (annotationClass in ModelViewProcessor.modelPropAnnotationsNames) continue`

juliafu1 commented:
@elihart all comments addressed. I also tested the parent view case (if the parent view has an attribute, the generated child view model shall have that attribute as well). The test is added in https://docs.google.com/document/d/1sT-AUi-Iiv3qI6dTfzbW3szRteq9qwaY-m1L7vnSY58/edit#bookmark=id.t7rle9y2y6xm. Thank you.

juliafu1 commented:
@elihart Tests added. PTAL. Thank you

elihart commented:
Great! tests look pretty exhaustive - I'll try to give this one final lookover soon and then merge 👍

juliafu1 commented:
Thank you!
